core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java

```diff
@@ -0,0 +1,275 @@
+   * a prefetch job was scheduled or not.
+   *
+   * @param path the Alluxio path
+   * @param mountTable the Alluxio mount table
+   * @return child UFS statuses of the alluxio path
+   * @throws InvalidPathException if the alluxio path can't be resolved to a UFS mount
+   * @see {@link #fetchChildrenIfAbsent(AlluxioURI, MountTable, boolean)}
+   */
+  public Collection<UfsStatus> fetchChildrenIfAbsent(AlluxioURI path, MountTable mountTable)
+      throws IOException, InvalidPathException {
+    return fetchChildrenIfAbsent(path, mountTable, true);
+  }
+
+  /**
+   * Retrieves the child UFS statuses for a given path and stores them in the cache.
+   *
+   * This method first checks if the children have already been retrieved, and if not, then
+   * retrieves them.
+
+   * @param path the path to get the children for
+   * @param mountTable the Alluxio mount table
+   * @return the child statuses that were stored in the cache, or null if the UFS couldn't list the
+   *         statuses
+   * @throws InvalidPathException when the table can't resolve the mount for the given URI
+   */
+  @Nullable
+  private Collection<UfsStatus> getChildrenIfAbsent(AlluxioURI path, MountTable mountTable)
+      throws InvalidPathException {
+    Collection<UfsStatus> children = getChildren(path);
+    if (children != null) {
+      return children;
+    }
+    MountTable.Resolution resolution = mountTable.resolve(path);
+    AlluxioURI ufsUri = resolution.getUri();
+    try (CloseableResource<UnderFileSystem> ufsResource = resolution.acquireUfsResource()) {
+      UnderFileSystem ufs = ufsResource.get();
+      UfsStatus[] statuses = ufs.listStatus(ufsUri.toString());
+      if (statuses == null) {
+        return null;
+      }
+      children = Arrays.asList(statuses);
+      addChildren(path, children);
+    } catch (IllegalArgumentException | IOException e) {
+      LOG.debug("Failed to add status to cache", e);
+    }
+    return children;
+  }
+
+  /**
+   * Get the child {@link UfsStatus}es from a given {@link AlluxioURI}.
+   *
+   * @param path the path the retrieve
+   * @return The corresponding {@link UfsStatus} or {@code null} if there is none stored
+   */
+  @Nullable
+  public Collection<UfsStatus> getChildren(AlluxioURI path) {
+    return mChildren.get(path);
+  }
+
+  /**
+   * Submit a request to asynchronously fetch the statuses corresponding to a given directory.
+   *
+   * Retrieve any fetched statuses by calling {@link #fetchChildrenIfAbsent(AlluxioURI, MountTable)}
+   * with the same Alluxio path.
+   *
+   * If no {@link ExecutorService} was provided to this object before instantiation, this method is
+   * a no-op.
+   *
+   * @param path the path to prefetch
+   * @param mountTable the Alluxio mount table
+   */
+  public void prefetchChildren(AlluxioURI path, MountTable mountTable) {
+    if (mPrefetchExecutor == null) {
+      return;
+    }
+    try {
+      Future<Collection<UfsStatus>> fute =
```

Comment:
hrmmm, is `fute` a real term? I am not familiar with it...

Reply:
updated to `job`